### PR TITLE
Fix inconsistency with Polars when doing quantile nearest on even length columns

### LINKED ISSUE #19324: [FEA] Allow controlling rounding behavior for `quantile(interp="nearest")` to match Polars
**Is your feature request related to a problem? Please describe.**
In polars, when using `quantile(interpolation="nearest")`, it rounds halfway values up. cudf currently rounds them down. This causes a mismatch when running polars with the GPU engine. For example,

```python
In [1]: import polars as pl

In [2]: pl.LazyFrame({"a": [1, 2, 3]}).select(pl.col("a").quantile(0.25, interpolation="nearest")).collect(engine=pl.GPUEngine(raise_on_fail=Tr
      ⋮ ue)) # GPU
Out[2]: 
shape: (1, 1)
┌─────┐
│ a   │
│ --- │
│ f64 │
╞═════╡
│ 1.0 │
└─────┘

In [3]: pl.LazyFrame({"a": [1, 2, 3]}).select(pl.col("a").quantile(0.25, interpolation="nearest")).collect() # CPU
Out[3]: 
shape: (1, 1)
┌─────┐
│ a   │
│ --- │
│ f64 │
╞═════╡
│ 2.0 │
└─────┘
```
**Describe the solution you'd like**
We would like to be able to choose how rounding is done for `quantile(interp="nearest")`, for example by passing a `rounding="half_up"` option to match Polars behavior.
**Describe alternatives you've considered**
come up with a workaround in cudf-polars

**Additional context**
Pylibcudf example
```python
import pylibcudf as plc

data = plc.Column.from_iterable_of_py([1, 2, 3], dtype=plc.DataType(plc.TypeId.INT64))

quarter = plc.reduce.reduce(
    data, 
    plc.aggregation.quantile(
        quantiles=[0.25],
        interp=plc.types.Interpolation.NEAREST,
        rounding=plc.round.RoundingMethod.HALF_UP # <-Add this
        ),
    plc.DataType(plc.TypeId.FLOAT64)
)
print(plc.interop.to_arrow(quarter))
```

### PR DESCRIPTION
**Pull request by jberg5**
## Description

Another fun one - basically Polars uses `f64::round()` which rounds 0.5 to 1, but cudf uses `cuda::std::nearbyint` which rounds 0.5 to 0. Illustrative example:

```python
    df = pl.LazyFrame(
        {
            "key": ["a", "a", "b", "b"],
            "val": [1.5, 2.5, 10.0, 20.0],
        }
    )
    q = df.group_by("key").agg(
        pl.col("val").quantile(0.5, interpolation="nearest").alias("q50")
    )
```

cpu returns [2.5, 20.0] (rounds up), gpu returns [1.5, 10.0] (rounds down).

Polars code:
https://github.com/pola-rs/polars/blob/8f2501b9cc5bba07a26b19594915c8976d7d5c9c/crates/polars-core/src/chunked_array/ops/aggregate/quantile.rs#L31

cudf:
https://github.com/rapidsai/cudf/blob/db19bac82e72f3f2666f3678751dfed0c21da6c2/cpp/src/quantiles/quantiles_util.hpp#L86

- Closes #19324 

## Checklist
- [x] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cudf/blob/HEAD/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.

### CODE REVIEW

**Review by Matt711 — CHANGES_REQUESTED** [review]
Thanks for the PR @jberg5!

We should not just blanket change the algorithm libcudf uses because it serves pandas in addition to polars. Did you check that the equivalent pandas MRE works correctly?

```python
import cudf.pandas
cudf.pandas.install()
import pandas as pd

df = pd.DataFrame({
    "key": ["a", "a", "b", "b"],
    "val": [1.5, 2.5, 10.0, 20.0],
})

q = df.groupby("key")["val"].quantile(0.5, interpolation="nearest").reset_index(name="q50")
print(q)
```

The fix should work for both libraries. See a proposed solution in #19324 for example.

Also, I noticed you've opened a few other PRs. In general, please open or check for an existing issue before opening a PR. It helps us with tracking. See the [contributing guide](https://github.com/rapidsai/cudf/blob/main/CONTRIBUTING.md). Thanks!

**Comment by jberg5** [comment]
Thanks for the feedback @Matt711 ! Good shout about pandas; I admit I've been narrowly focused on polars and didn't test the equivalent pandas case. I think your proposed solution is a good one. I took a stab at implementing that approach here, lmk if it matches what you had in mind.

Apologies for missing the existing issue, and for not including an issue on my other PRs! Will do so in the future.

**Review by davidwendt — CHANGES_REQUESTED** [review]
Missing libcudf gtests
https://github.com/rapidsai/cudf/blob/585fbcad3ba004bbde4836c9f5b762f343e5ec05/cpp/tests/quantiles/quantile_test.cpp#L378

File: cpp/src/quantiles/quantiles_util.hpp [hunk 1]
```diff
@@ -74,17 +74,19 @@ struct quantile_index {
...
   {
     quantile = cuda::std::min(cuda::std::max(quantile, 0.0), 1.0);
 
-    double val = quantile * (count - 1);
-    lower      = std::floor(val);
-    higher     = static_cast<size_type>(cuda::std::ceil(val));
-    nearest    = static_cast<size_type>(cuda::std::nearbyint(val));
-    fraction   = val - lower;
+    double val      = quantile * (count - 1);
+    lower           = std::floor(val);
```
Comment (wence-) [inline cpp/src/quantiles/quantiles_util.hpp]: *nit*: I know this is just a copy of existing code, but we should use `cuda::std::floor` here.

**Review by wence- — CHANGES_REQUESTED** [review]

**Comment by jberg5** [comment]
Have since added gtests and switched to `cuda::std::floor`, thanks @wence and @davidwendt, lmk if this looks good to you.

**Comment by davidwendt** [comment]
/ok to test https://github.com/rapidsai/cudf/pull/21708/commits/27c98da2304a1f92daaa1aea385577f0a424a584

**Comment by davidwendt** [comment]
/ok to test https://github.com/rapidsai/cudf/pull/21708/commits/59ebac49e47bed65545725b83632e318c3a43cf6

File: python/cudf_polars/tests/test_groupby.py [hunk 2]
```diff
@@ -459,6 +459,19 @@ def test_groupby_sum_decimal_null_group() -> None:
     assert_gpu_result_equal(q, check_row_order=False)
 
 
+def test_groupby_quantile_nearest_even_length():
```
Comment (Matt711) [inline python/cudf_polars/tests/test_groupby.py]: The default behavior matched pandas. Did you check if this is a redundant test?

**Review by Matt711 — CHANGES_REQUESTED** [review]

File: python/cudf_polars/tests/test_groupby.py (hunk 2 above)
Comment (jberg5) [reply python/cudf_polars/tests/test_groupby.py]: Good q, so the closest existing test I found was for https://github.com/rapidsai/cudf/blob/6c16c8aa1b4c0c135b46744f944be06ce26b13d9/python/cudf_polars/tests/test_groupby.py#L96 But it doesn't catch this edge case specifically.

If you'd like I can add `[pl.col("float").quantile(0.5, interpolation="nearest")]` to _EXPRS, this would be much less code and I'm pretty sure it would prevent a regression on this specific edge case, but it wouldn't be super obvious to future people what it was there for.

**Comment by Matt711** [comment]
/ok to test https://github.com/rapidsai/cudf/pull/21708/commits/d98743b54b5ae2cbe2bb8d18150ed98674fa0e90

**Comment by Matt711** [comment]
/ok to test https://github.com/rapidsai/cudf/pull/21708/commits/a99d7d5736c46bae3a06ce5be34c68632538233c

**Comment by vyasr** [comment]
/ok to test https://github.com/rapidsai/cudf/pull/21708/commits/152155be5cd3b8c663bb400196f57177225782d8

**Review by davidwendt — APPROVED** [review]

File: python/cudf_polars/tests/test_groupby.py [hunk 3]
```diff
@@ -710,6 +710,19 @@ def test_groupby_sum_decimal_null_group(
...
+    df = pl.LazyFrame(
+        {
+            "key": ["a", "a", "b", "b"],
+            "val": [1.5, 2.5, 10.0, 20.0],
+        }
+    )
+    q = df.group_by("key").agg(
+        pl.col("val").quantile(0.5, interpolation="nearest").alias("q50")
+    )
+    assert_gpu_result_equal(q, check_row_order=False)
```
Comment (wence-) [inline python/cudf_polars/tests/test_groupby.py]: Can we add a test that nearest_half_up rounds away from zero as well. Presumably the  also `(-pl.col("val")).quantile(...)` would do the trick?

**Review by wence- — APPROVED** [review]
Approving with one request to add a test that half_up interpolation rounds away from zero.

**Comment by Matt711** [comment]
/ok to test https://github.com/rapidsai/cudf/pull/21708/commits/19a2ddd5e95d9ceb8b7a7ddd5403c00bdea982ea

**Review by Matt711 — APPROVED** [review]
Thanks!
